### PR TITLE
[perf] skip the per-forward RoPE buffer broadcast in the RL trainers

### PR DESCRIPTION
**Pull request by basil-k-aji-dev**
Follow-up to your review on #854. Thanks for the two-rank revalidation — it prompted me to test these the same way rather than reason about them.

This takes two of your three suggestions. The third I'd rather not take as-is, and I have a reproduction showing why.

## 1. `broadcast_buffers=False`

DDP re-broadcasts every registered buffer from rank 0 on each forward. The model registers exactly two:

```python
self.register_buffer("freqs_cos", freqs_cos, persistent=False)
self.register_buffer("freqs_sin", freqs_sin, persistent=False)
```

Both are pure functions of config (`head_dim`, `max_position_embeddings`, `rope_theta`, `rope_scaling`), never data-dependent, and the one place they are reassigned — the meta-device recovery path in `MiniMindModel.forward` — recomputes them from the same config, so it is deterministic across ranks too. The broadcast moves identical bytes every step.

Applied to `train_ppo.py` (actor and critic), `train_grpo.py` and `train_agent.py`.

## 2. The critic's tied-embedding assumption

Documented on `CriticModel`. The reason this class works under DDP at all is subtle: `forward` never calls `lm_head`, but `tie_word_embeddings` (default `True`) makes

```python
self.model.embed_tokens.weight = self.lm_head.weight
```

the *same* Parameter, and the embedding is used on every forward — so the tied parameter still gets a gradient and DDP sees no unused parameters. Untie it and `lm_head.weight` becomes a real parameter this forward never touches, and DDP raises `Expected to have finished reduction in the prior iteration` unless the critic is wrapped with `find_unused_parameters=True` or `lm_head` is deleted.

## 3. `no_sync()` — not taken, and here is the failure

The naive version is unsafe in all three trainers, because each one ends with a trailing flush:

```python
if grad_accum_step % args.accumulation_steps != 0:
    ...
    actor_optimizer.step()
```

If the non-boundary micro-batches ran under `no_sync()`, the gradients that flush steps on were only ever accumulated locally and never all-reduced — silently reintroducing the divergence #854 fixed. PPO makes it likelier, since `stop_ppo` can break the outer loop mid-window.

Reproduced on two CPU/Gloo ranks (`accumulation_steps=2`, three micro-batches, so the loop ends on a non-boundary step):

```text
PASS  bb_false_grads_match             = True
PASS  bb_false_buffers_match           = True
FAIL  naive_nosync_tail_grads_match    = False
```

Doing it correctly needs the sync to be forced on whichever backward precedes an optimiser step — including the final micro-batch when the trailing flush will fire, and when `stop_ppo` is set. Correct use also requires wrapping the **forward** as well as the backward, which in `train_grpo.py` spans ~30 lines between the forward and `loss.backward()`.

Worth noting the payoff is currently zero anyway: `accumulation_steps` defaults to `1` in all three trainers, so every step is a boundary step and `no_sync()` never engages.

Happy to send it as its own PR with the guard plus the Gloo test if you want it — it seemed better to keep a correctness-sensitive change out of a patch that is otherwise trivially safe.

## Verification

Two-rank CPU/Gloo, each rank fed different data so unsynchronised gradients would differ, then `all_gather` compared:

- gradients identical across ranks with `broadcast_buffers=False`
- RoPE buffers identical across ranks without the broadcast
- `python -m py_compile` clean on all three trainers

One note for later: torch 2.14 deprecates `broadcast_buffers` in favour of `forward_sync_buffers`. I kept `broadcast_buffers`, since `requirements.txt` references torch 2.6 where the new name does not exist; on 2.14+ it still works but emits a `FutureWarning`.

### CODE REVIEW

**Review by alikartaraa-web — COMMENTED** [review]
-
